### PR TITLE
#161290920 Role Based Functionality

### PR DESCRIPTION
**Pull request by oluseyianani**
#### What does this PR do?
- Specifies roles available in Authors Haven (Users, Admins, and Super Admins)
- Implements functionality to create the super admin
- Implements functionality for super admin to assign admins by specifying their username
- Allows admin to be able to assign other admins
- Allows Super Admin be able to revoke admin right and privileges. 
- Allows Admins and the super admin to be able to resolve and, or block user's comments and articles on Authors Haven
- Gives users the right to report offensive Articles, and comments.
- Enable Admins and Super admin take down User's account
- Enables Admins and Super admin reactivate user's account
- Enable Admin and Super Admin to add new categories to Authors Haven.

### Roles
- Assign SuperAdmin `POST /api/v1/role/superadmin/assign  /200 OK`
- Super admin assign Admins - `PUT /api/v1/role/admin/assign  /200 OK`
`--Required field: username`
- Super Admin Revoke admin's priviledges - `PUT api/v1/role/admin/unassign     /200 OK`

### Articles
- Take down an article `DELETE /api/v1/admin/takedown/articles/:slug`
`---- where :slug is article slug`
- Add new Article Categories `POST /api/v1/categories`

### Users
- Deactivate Users `DELETE /api/v1/admin/deactivate/users/:username  /204 NO CONTENT`
- Reactivate Users `PUT /api/v1/admin/reactivate/users/:username /200 OK`

### Comments
- Report Comments `POST /api/articles/:articleId/comments/:commentId/report /200 OK`
`-- Required fields: reportType  and reportDetail`
- Admin or SuperAdmin Resolve comment complaint `PUT /api/articles/:articleId/comments/:commentId/resolve`
`--Required field: decision`
`--Required values: blocked or resolved`
- Users can edit their reported comment `PUT /api/v1/articles/:articleId/comments/:commentId/resolve/editcomments`
`-- Required field: acceptedComment`
- Admin and SuperAdmin can Sort reported comments by Status: pending, blocked, or resolved.
`GET /api/v1/articles/comments/reports?status=pending`
`GET /api/v1/articles/comments/reports?status=blocked`
`GET /api/v1/articles/comments/reports?status=resolved`
`--Required field as a query: status`

#### How should this be manually tested?
- After cloning this repo, `cd` into it and:
Create users by following this instruction #11 
Create Articles by following this instruction #13 
Create Comments by following this intruction #35 

- To assign Super Admin, 
Set the Email address, Usernme and Password on the environment with 
Example:
SUPER_USER_EMAIL= emailHere@email.com
SUPER_USER_PASSWORD= passwordHere
SUPER_USERNAME= superAdminUsername

Once that is set, hit the route;
POST `/api/v1/role/superadmin/assign`

- To assign Admins
Login the Super admin with the credentials you've provided.
Get the username of the user you want to assign an admin priviledge to.
Once that is done, hit this route:
PUT`/api/v1/role/admin/assign`

- To revoke admin priviledges
Get the admin's username and hit the route:
`PUT /api/v1/role/admin/unassign`

- To take down an article, get the article slug and use this route:
DELETE `/api/v1/admin/takedown/articles/:slug`

- To add new Article Categories 
use this route POST `/api/v1/categories`

- To deactivate a user, get the username and use the route: 
DELETE `/api/v1/admin/deactivate/users/:username`

- To Reactivate a user, get the username and use the route:  
PUT `/api/v1/admin/reactivate/users/:username`

- To report a comment, get the article id and the comment id, then his this route
POST `/api/articles/:articleId/comments/:commentId/report`
`-- Required fields: reportType  and reportDetail`

- For a user to edit their reported comment, they need to be logged in and hit this route
PUT `/api/v1/articles/:articleId/comments/:commentId/resolve/editcomments`
- Required field is acceptedComment

- For admins or super admins to resolve reported comments
Go to: PUT `/api/articles/:articleId/comments/:commentId/resolve`
- Required fields: decision
- Required values: resolved or blocked

- For admins or super admins to sort status by either pending, resolved or blocked
Admins or Super admins should visit:
`GET /api/v1/articles/comments/reports?status=pending`
`GET /api/v1/articles/comments/reports?status=blocked`
`GET /api/v1/articles/comments/reports?status=resolved`

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
#161290920
#### Screenshots (if appropriate)
screen shot 2018-11-22 at 12 27 25 pm

screen shot 2018-11-22 at 12 36 48 pm

screen shot 2018-11-22 at 12 39 48 pm

takedown

add categories

deactivate a user

reactivate a user

deactivateemail

reactivateemail

article takedown

screen shot 2018-11-22 at 1 01 10 pm
screen shot 2018-11-22 at 1 05 25 pm

screen shot 2018-11-22 at 1 08 55 pm

screen shot 2018-11-22 at 1 14 22 pm

screen shot 2018-11-22 at 1 18 19 pm

screen shot 2018-11-22 at 1 19 55 pm

#### Questions:
N/A

### CODE REVIEW

**Review by ovieokeh — APPROVED** [review]
Awesome work for getting this done and dusted @anasey01 !!

**Review by desaawa — COMMENTED** [review]
Please resolve conflicts

**Comment by hope-uwa** [comment]
@anasey01 I noticed the logs/info.log is not ignored, it displays private information like users details. Please add it to the .gitignore file.

**Review by oluseyianani — COMMENTED** [review]
@uwaelpis Thank you for the feedback. However, the log file is dependent on a unit test that was written by @ajimae. Ignoring those files will mean failing tests which means we have to figure out a work around.

File: logs/info.log [hunk 1]
```diff
@@ -83,3 +83,211 @@
    { from: 'no-reply@authorshaven.com',
      to: [ 'meeky.ae@gmail.com' ] },
   messageId: '<8aa8fa41-2558-3751-b3d5-a6af9b231022@authorshaven.com>' }
+[Fri Nov 23 2018 12:39:42 GMT+0100 (West Africa Standard Time)] : { accepted: [ 'uwa@noldor.com' ],
```
Comment (desaawa) [inline logs/info.log:86]: you need to find a work around to remove this.

**Review by desaawa — APPROVED** [review]

**Pull request merged**